In [431]:
import numpy as np
import pandas as pd
from pathlib import Path
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.datasets import make_classification
from sklearn.ensemble import RandomForestClassifier

In [432]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))
train_df

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0           57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1          141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2          321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3           11778   11778     3000.0    0.1240       100.22           RENT   
4          169382  169382    30000.0    0.1612      1056.49       MORTGAGE   
...           ...     ...        ...       ...          ...            ...   
12175      354912  354912    19975.0    0.2565       801.09           RENT   
12176      354944  354944    15000.0    0.1774       540.34           RENT   
12177      354973  354973     3600.0    0.1862       131.28           RENT   
12178      355002  355002    15000.0    0.0881       475.68       MORTGAGE   
12179      355307  355307    15000.0    0.1774       540.34           RENT   

       annual_inc verification_status loan_status pymnt_plan  ...  \
0        223000.0        Not Verified    low_risk          n  ...   
1        123000.0     Source Verified    low_risk          n  ...   
2        197000.0     Source Verified    low_risk          n  ...   
3         45000.0        Not Verified    low_risk          n  ...   
4        133000.0     Source Verified    low_risk          n  ...   
...           ...                 ...         ...        ...  ...   
12175     28000.0        Not Verified   high_risk          n  ...   
12176     50000.0            Verified   high_risk          n  ...   
12177     60000.0        Not Verified   high_risk          n  ...   
12178     62000.0     Source Verified   high_risk          n  ...   
12179     54080.0        Not Verified   high_risk          n  ...   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0               100.0              50.0                   0.0        0.0   
1                85.0              33.3                   0.0        0.0   
2                85.7              33.3                   0.0        0.0   
3               100.0              16.7                   1.0        0.0   
4               100.0              66.7                   0.0        0.0   
...               ...               ...                   ...        ...   
12175           100.0              16.7                   0.0        0.0   
12176            90.5              11.1                   0.0        0.0   
12177           100.0               0.0                   0.0        0.0   
12178           100.0               0.0                   0.0        0.0   
12179            90.0             100.0                   0.0        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0             577150.0           122018.0         32000.0   
1             132750.0            27896.0         15900.0   
2             628160.0           114043.0         22600.0   
3              42006.0            20761.0         19900.0   
4             283248.0           109056.0         79500.0   
...                ...                ...             ...   
12175          50055.0            28192.0         18700.0   
12176          70324.0            57025.0         13300.0   
12177          83765.0            55156.0         14800.0   
12178         189930.0            23748.0          7000.0   
12179          48968.0            29786.0          2900.0   

      total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                       170200.0              N                     N  
1                        35398.0              N                     N  
2                        90340.0              N                     N  
3                        15406.0              N                     N  
4                        58778.0              N                     N  
...                          ...            ...                   ...  
12175                    19055.0              N                     N  
12176              

In [433]:
 #Checking for missing values
test_df.isnull().sum()

Unnamed: 0                    0
index                         0
loan_amnt                     0
int_rate                      0
installment                   0
                             ..
total_bal_ex_mort             0
total_bc_limit                0
total_il_high_credit_limit    0
hardship_flag                 0
debt_settlement_flag          0
Length: 86, dtype: int64

In [434]:
#drop the label to create the X data
X = train_df.drop('index', axis=1)
X


Unnamed: 0  loan_amnt  int_rate  installment home_ownership  \
0           57107    13375.0    0.1797       483.34       MORTGAGE   
1          141451    21000.0    0.1308       478.68       MORTGAGE   
2          321143    20000.0    0.1240       448.95       MORTGAGE   
3           11778     3000.0    0.1240       100.22           RENT   
4          169382    30000.0    0.1612      1056.49       MORTGAGE   
...           ...        ...       ...          ...            ...   
12175      354912    19975.0    0.2565       801.09           RENT   
12176      354944    15000.0    0.1774       540.34           RENT   
12177      354973     3600.0    0.1862       131.28           RENT   
12178      355002    15000.0    0.0881       475.68       MORTGAGE   
12179      355307    15000.0    0.1774       540.34           RENT   

       annual_inc verification_status loan_status pymnt_plan    dti  ...  \
0        223000.0        Not Verified    low_risk          n  29.99  ...   
1        123000.0     Source Verified    low_risk          n  11.26  ...   
2        197000.0     Source Verified    low_risk          n  11.28  ...   
3         45000.0        Not Verified    low_risk          n  18.08  ...   
4        133000.0     Source Verified    low_risk          n  27.77  ...   
...           ...                 ...         ...        ...    ...  ...   
12175     28000.0        Not Verified   high_risk          n  28.42  ...   
12176     50000.0            Verified   high_risk          n  23.43  ...   
12177     60000.0        Not Verified   high_risk          n  28.80  ...   
12178     62000.0     Source Verified   high_risk          n  11.44  ...   
12179     54080.0        Not Verified   high_risk          n  21.66  ...   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0               100.0              50.0                   0.0        0.0   
1                85.0              33.3                   0.0        0.0   
2                85.7              33.3                   0.0        0.0   
3               100.0              16.7                   1.0        0.0   
4               100.0              66.7                   0.0        0.0   
...               ...               ...                   ...        ...   
12175           100.0              16.7                   0.0        0.0   
12176            90.5              11.1                   0.0        0.0   
12177           100.0               0.0                   0.0        0.0   
12178           100.0               0.0                   0.0        0.0   
12179            90.0             100.0                   0.0        0.0   

       tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0             577150.0           122018.0        32000.0   
1             132750.0            27896.0        15900.0   
2             628160.0           114043.0        22600.0   
3              42006.0            20761.0        19900.0   
4             283248.0           109056.0        79500.0   
...                ...                ...            ...   
12175          50055.0            28192.0        18700.0   
12176          70324.0            57025.0        13300.0   
12177          83765.0            55156.0        14800.0   
12178         189930.0            23748.0         7000.0   
12179          48968.0            29786.0         2900.0   

       total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                        170200.0              N                     N  
1                         35398.0              N                     N  
2                         90340.0              N                     N  
3                         15406.0              N                     N  
4                         58778.0              N                     N  
...                           ...            ...                   ...  
12175                     19055.0              N                     N  
12176                     54824.0  

In [435]:
X_dummies = pd.get_dummies(X)
print(X_dummies.columns)
X_dummies

Index(['Unnamed: 0', 'loan_amnt', 'int_rate', 'installment', 'annual_inc',
       'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec',
       'revol_bal', 'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_s

Unnamed: 0  loan_amnt  int_rate  installment  annual_inc    dti  \
0           57107    13375.0    0.1797       483.34    223000.0  29.99   
1          141451    21000.0    0.1308       478.68    123000.0  11.26   
2          321143    20000.0    0.1240       448.95    197000.0  11.28   
3           11778     3000.0    0.1240       100.22     45000.0  18.08   
4          169382    30000.0    0.1612      1056.49    133000.0  27.77   
...           ...        ...       ...          ...         ...    ...   
12175      354912    19975.0    0.2565       801.09     28000.0  28.42   
12176      354944    15000.0    0.1774       540.34     50000.0  23.43   
12177      354973     3600.0    0.1862       131.28     60000.0  28.80   
12178      355002    15000.0    0.0881       475.68     62000.0  11.44   
12179      355307    15000.0    0.1774       540.34     54080.0  21.66   

       delinq_2yrs  inq_last_6mths  open_acc  pub_rec  ...  \
0              0.0             0.0      15.0      0.0  ...   
1              2.0             0.0      16.0      0.0  ...   
2              0.0             0.0      12.0      0.0  ...   
3              0.0             0.0      12.0      1.0  ...   
4              0.0             2.0      13.0      0.0  ...   
...            ...             ...       ...      ...  ...   
12175          0.0             0.0      15.0      0.0  ...   
12176          4.0             0.0      16.0      0.0  ...   
12177          0.0             1.0      14.0      0.0  ...   
12178          0.0             0.0       5.0      0.0  ...   
12179          1.0             1.0       5.0      0.0  ...   

       loan_status_low_risk  pymnt_plan_n  initial_list_status_f  \
0                         1             1                      0   
1                         1             1                      0   
2                         1             1                      0   
3                         1             1                      0   
4                         1             1                      0   
...                     ...           ...                    ...   
12175                     0             1                      0   
12176                     0             1                      0   
12177                     0             1                      0   
12178                     0             1                      0   
12179                     0             1                      1   

       initial_list_status_w  application_type_Individual  \
0                          1                            1   
1                          1                            1   
2                          1                            1   
3                          1                            1   
4                          1                            1   
...                      ...                          ...   
12175                      1                            1   
12176                      1                            1   
12177                      1                            1   
12178                      1                            0   
12179                      0                            1   

       application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                               0                1                0   
1                               0                1                0   
2                               0                1                0   
3                               0                1                0   
4                               0                1                0   
...                           ...              ...              ...   
12175                           0                1                0   
12176                           0                1                0   
12177                           0                1                0   
12178                           1                1                0   
12179                           0         

In [436]:
# Convert categorical data to numeric and separate target feature for training data
y_train = train_df['loan_status']
X_train = train_df.drop(columns = ["loan_status"])
X_train = pd.get_dummies(X_train)
X_train


Unnamed: 0   index  loan_amnt  int_rate  installment  annual_inc  \
0           57107   57107    13375.0    0.1797       483.34    223000.0   
1          141451  141451    21000.0    0.1308       478.68    123000.0   
2          321143  321143    20000.0    0.1240       448.95    197000.0   
3           11778   11778     3000.0    0.1240       100.22     45000.0   
4          169382  169382    30000.0    0.1612      1056.49    133000.0   
...           ...     ...        ...       ...          ...         ...   
12175      354912  354912    19975.0    0.2565       801.09     28000.0   
12176      354944  354944    15000.0    0.1774       540.34     50000.0   
12177      354973  354973     3600.0    0.1862       131.28     60000.0   
12178      355002  355002    15000.0    0.0881       475.68     62000.0   
12179      355307  355307    15000.0    0.1774       540.34     54080.0   

         dti  delinq_2yrs  inq_last_6mths  open_acc  ...  \
0      29.99          0.0             0.0      15.0  ...   
1      11.26          2.0             0.0      16.0  ...   
2      11.28          0.0             0.0      12.0  ...   
3      18.08          0.0             0.0      12.0  ...   
4      27.77          0.0             2.0      13.0  ...   
...      ...          ...             ...       ...  ...   
12175  28.42          0.0             0.0      15.0  ...   
12176  23.43          4.0             0.0      16.0  ...   
12177  28.80          0.0             1.0      14.0  ...   
12178  11.44          0.0             0.0       5.0  ...   
12179  21.66          1.0             1.0       5.0  ...   

       verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                                 0             1                      0   
1                                 0             1                      0   
2                                 0             1                      0   
3                                 0             1                      0   
4                                 0             1                      0   
...                             ...           ...                    ...   
12175                             0             1                      0   
12176                             1             1                      0   
12177                             0             1                      0   
12178                             0             1                      0   
12179                             0             1                      1   

       initial_list_status_w  application_type_Individual  \
0                          1                            1   
1                          1                            1   
2                          1                            1   
3                          1                            1   
4                          1                            1   
...                      ...                          ...   
12175                      1                            1   
12176                      1                            1   
12177                      1                            1   
12178                      1                            0   
12179                      0                            1   

       application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                               0                1                0   
1                               0                1                0   
2                               0                1                0   
3                               0                1                0   
4                               0                1                0   
...                           ...              ...              ...   
12175                           0                1                0   
12176                           0                1                0   
12177                           0                1                0   
12178                        

In [437]:
# Convert categorical data to numeric and separate target feature for testing data
y_test = test_df['loan_status']
X_test = test_df.drop(columns = ["loan_status"])
X_test = pd.get_dummies(X_test)
# X_dummies
X_test

Unnamed: 0  index  loan_amnt  int_rate  installment  annual_inc    dti  \
0          67991  67991    40000.0    0.0819       814.70    140000.0  19.75   
1          25429  25429     6000.0    0.1524       208.70     55000.0  11.52   
2          38496  38496     3600.0    0.1695       128.27     42000.0   6.74   
3          19667  19667    20000.0    0.1524       478.33    100000.0  12.13   
4          37505  37505     3600.0    0.1240       120.27     50000.0  16.08   
...          ...    ...        ...       ...          ...         ...    ...   
4697       77282  77282    30000.0    0.1240       673.42    140480.0  15.74   
4698       77291  77291    24000.0    0.0756       747.22     50000.0  26.81   
4699       77292  77292    10000.0    0.2305       387.36     33000.0  38.51   
4700       77297  77297     8000.0    0.1862       205.86     38000.0  16.36   
4701       77304  77304    30000.0    0.2055      1123.34    180000.0  12.06   

      delinq_2yrs  inq_last_6mths  open_acc  ...  \
0             0.0             1.0      18.0  ...   
1             2.0             0.0       8.0  ...   
2             0.0             0.0       6.0  ...   
3             0.0             2.0       7.0  ...   
4             0.0             3.0       6.0  ...   
...           ...             ...       ...  ...   
4697          0.0             0.0      20.0  ...   
4698          0.0             0.0       9.0  ...   
4699          0.0             2.0       7.0  ...   
4700          0.0             1.0       8.0  ...   
4701          0.0             0.0       8.0  ...   

      verification_status_Source Verified  verification_status_Verified  \
0                                       0                             0   
1                                       0                             0   
2                                       0                             0   
3                                       0                             0   
4                                       0                             0   
...                                   ...                           ...   
4697                                    1                             0   
4698                                    0                             0   
4699                                    0                             1   
4700                                    1                             0   
4701                                    1                             0   

      pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0                1                      0                      1   
1                1                      0                      1   
2                1                      0                      1   
3                1                      0                      1   
4                1                      0                      1   
...            ...                    ...                    ...   
4697             1                      1                      0   
4698             1                      0                      1   
4699             1                      1                      0   
4700             1                      0                      1   
4701             1                      1                      0   

      application_type_Individual  application_type_Joint App  \
0                               1                           0   
1                               1                           0   
2                               1                           0   
3                               1                           0   
4                               1                           0   
...                           ...                         ...   
4697                            1                           0   
4698                            1                           0   
4699                            1                           0   
4700                            1                   

In [438]:
# add missing dummy variables to testing set
X_test = X_test.reindex(columns = X_train.columns, fill_value=0)

In [439]:
# Train the Logistic Regression model on the unscaled data and print the model score

clf = LogisticRegression().fit(X_train, y_train)
print(f'Training Score: {clf.score(X_train, y_train)}')
print(f'Testing Score: {clf.score(X_test, y_test)}')

Training Score: 0.6485221674876848
Testing Score: 0.5253083794130158


/Users/mindy/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [440]:
# Train a Random Forest Classifier model and print the model score
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train, y_train)
print(f'Training Score: {clf.score(X_train, y_train)}')
print(f'Testing Score: {clf.score(X_test, y_test)}')

Training Score: 1.0
Testing Score: 0.6180348787749894


In [441]:
# Scale the data
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [442]:
# Train the Logistic Regression model on the scaled data and print the model score
log_scaled = LogisticRegression().fit(X_train_scaled, y_train)
print(f'Training Score: {log_scaled.score(X_train_scaled, y_train)}')
print(f'Testing Score: {log_scaled.score(X_test_scaled, y_test)}')

Training Score: 0.713136288998358
Testing Score: 0.7201190982560612


/Users/mindy/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [443]:
# Train a Random Forest Classifier model on the scaled data and print the model score
forest_scaled = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train, y_train)
print(f'Training Score: {forest_scaled.score(X_train, y_train)}')
print(f'Testing Score: {forest_scaled.score(X_test, y_test)}')

Training Score: 1.0
Testing Score: 0.6180348787749894


    I did not expect the data from Random Forest classifier to be affected by scaling because scaling is not necessary for random 
    forests due to it being a tree based model. As you can see, the scaled data was the same for both scaled and unscaled data. 

    The machine learning algorithms, random forest, and logistic regression are utilized for various classification and regression purposes. In this case, the Random Forest classifier seems to outperform Logistic Regression; however, the difference is not immense.

    Results:

    Logistic Regression on unscaled data: 0.5253083794130158
    Random Forest Classifier on unscaled data: 0.6180348787749894

    Logistic Regression on scaled data: 0.7201190982560612
    Random Forest Classifier on scaled data: 0.6180348787749894




    

